addon/components/animated-popper.js

```diff
@@ -0,0 +1,197 @@
+import Component from '@ember/component';
+import { run } from '@ember/runloop';
+import { addObserver, removeObserver } from '@ember/object/observers';
+
+import { action } from 'ember-decorators/object';
+import { tagName } from 'ember-decorators/component';
+
+import { argument, type } from 'ember-argument-decorators';
+import { unionOf } from 'ember-argument-decorators/types';
+
+import { scheduler as raf, Token } from 'ember-raf-scheduler';
+
+import layout from '../templates/components/animated-popper';
+
+function hasTransition(element) {
+  const { transitionDuration } = window.getComputedStyle(element);
+
+  return parseFloat(transitionDuration) > 0;
+}
+
+/**
+ * Private component used by ice-pop internally. This component handles the logic
+ * surrounding animation, including determining if animations should happen at all.
+ */
+@tagName('')
+export default class AnimatedPopperComponent extends Component {
+  layout = layout;
+
+  // ----- Arguments ------
+
+  /**
+   * Whether or not the popper is open
+   */
+  @argument
+  @type('boolean')
+  isOpen = false;
+
+  /**
+   * Action sent when the popper has been added to the DOM, before animations
+   */
+  @argument
+  @type(unionOf(null, 'action'))
+  onOpen = null;
+
+  /**
+   * Action sent when the popper has been removed from the DOM, after animations
+   */
+  @argument
+  @type(unionOf(null, 'action'))
+  onClose = null;
+
+  /**
+   * Whether or not the popper should render in place or at the root level
+   */
+  @argument
+  @type('boolean')
+  renderInPlace = false;
+
+  /**
+   * The popper elment's class
+   */
+  @argument
+  @type('string')
+  popperClass = '';
+
+  /**
+   * The placement for the popper element
+   */
+  @argument
+  @type('string')
+  placement = 'auto';
+
+  // ----- Private Variables -----
+
+  /**
+   * Whether or not the popper should be rendered in the DOM
+   */
+  @type('boolean')
+  renderInDOM = false;
+
+  /**
+   * Wether or not the popper should have the `is-visible` class, which will trigger animations
+   */
+  @type('boolean')
+  makeVisible = false;
+
+  /**
+   * RAF scheduler token, used to cancel all active jobs
+   */
+  _token = new Token();
+
+  /**
+   * Empty text node used to find the true parent element for this component
+   */
+  _parentFinder = document.createTextNode('');
+
+  /**
+   * The popper element, stored in order to add and remove event handlers for transitions
+   */
+  _animatedElement = null;
+
+  /**
+   * Whether or not the current popper has a transition duration > 0
+   */
+  _hasTransition = false;
+
+  constructor() {
+    super(...arguments);
+
+    this.isOpenDidChange();
+  }
+
+  didInsertElement() {
+    this._target = this._parentFinder.parentNode.parentNode;
```

Comment:
Might be I am missing something but where `_target` is used in this file? And since `_parentFinder` is a constant, does it mean `_target` is also a constant?

Reply:
https://github.com/Addepar/ice-pop/pull/24/files#diff-cbc68c046ca4db42c5991993e709aa3aR5

Reply:
Added the field to the class with a comment describing its function